src/main/java/bio/terra/service/snapshotbuilder/utils/HierarchyQueryBuilder.java

```diff
@@ -92,73 +99,69 @@ public Query generateQuery(SnapshotBuilderDomainOption domainOption, int concept
-        BooleanAndOrFilterVariable.and(
-            BinaryFilterVariable.equals(
-                conceptAncestor.makeFieldVariable(ConceptAncestor.DESCENDANT_CONCEPT_ID),
-                conceptIdLiteral),
-            BinaryFilterVariable.notEquals(ancestorConceptId, conceptIdLiteral)));
+  static SourceVariable joinToFilterConcepts(FieldVariable parentId, int conceptId) {
+    var conceptAncestorTable =
+        SourceVariable.forPrimary(TablePointer.fromTableName(ConceptAncestor.TABLE_NAME));
+    var ancestorConceptId =
+        conceptAncestorTable.makeFieldVariable(ConceptAncestor.ANCESTOR_CONCEPT_ID);
+    var descendantConceptId =
+        conceptAncestorTable.makeFieldVariable(ConceptAncestor.DESCENDANT_CONCEPT_ID);
+    var subquery =
+        new Query(
+            List.of(ancestorConceptId, descendantConceptId),
+            List.of(conceptAncestorTable),
+            BooleanAndOrFilterVariable.and(
+                BinaryFilterVariable.equals(descendantConceptId, new Literal(conceptId)),
+                BinaryFilterVariable.notEquals(ancestorConceptId, new Literal(conceptId))),
+            null);
+    var subQueryPointer = new SubQueryPointer(subquery, "join_filter");
+    return SourceVariable.forJoined(subQueryPointer, ConceptAncestor.ANCESTOR_CONCEPT_ID, parentId);
   }
 
-  /**
-   * Generate a subquery that returns true if the outerConcept has any children.
-   *
-   * <pre>{@code
-   * EXISTS (SELECT
-   *     1
-   *   FROM
-   *     concept_ancestor ca
-   *   JOIN
-   *     concept c2
-   *   ON
-   *     c2.concept_id = ca.descendant_concept_id
-   *     AND ca.ancestor_concept_id = c.concept_id
-   *     AND ca.descendant_concept_id != c.concept_id
-   *     AND c2.standard_concept = 'S') AS has_children
-   * }</pre>
-   */
-  static SelectExpression hasChildrenExpression(FieldVariable conceptId) {
-    var conceptAncestor =
-        TableVariable.forPrimary(TablePointer.fromTableName(ConceptAncestor.TABLE_NAME));
+  static FieldVariable selectHChildren(SourceVariable joinHasChildren) {
+    return joinHasChildren.makeFieldVariable(
+        ConceptAncestor.DESCENDANT_CONCEPT_ID, "COUNT", "has_children", "> 0", true);
+  }
+
+  /** Generate a join clause that is used to determine if the outerConcept has any children. */
+  static SourceVariable makeHasChildrenJoin(FieldVariable childId) {
+    var conceptAncestorTable =
+        SourceVariable.forPrimary(TablePointer.fromTableName(ConceptAncestor.TABLE_NAME));
+    var ancestorConceptId =
+        conceptAncestorTable.makeFieldVariable(ConceptAncestor.ANCESTOR_CONCEPT_ID);
     var descendantConceptId =
-        conceptAncestor.makeFieldVariable(ConceptAncestor.DESCENDANT_CONCEPT_ID);
-    var innerConcept =
-        TableVariable.forJoined(
+        conceptAncestorTable.makeFieldVariable(ConceptAncestor.DESCENDANT_CONCEPT_ID);
+    var minLevelsSeparation =
+        conceptAncestorTable.makeFieldVariable(ConceptAncestor.MIN_LEVELS_OF_SEPARATION);
```

Comment:
If there are any children, at any level, don't we want to include them for the purposes of hasChildren? Does the min level of separation matter?

Reply:
Oh this is an optimization so we don;t need to compute such large values. Thanks for explaining that!